containership/e2e-test
# Implement initial PoC

Here's an initial PoC implementation to kick things off. I have some more experiments locally, but this is boiled down enough for a high-level review. There are a few things that should definitely be updated before merging which are elaborated on below.

Tests should be run as follows:

1. Provision:

```
ginkgo -r -v -debug -failFast provision -- --template `pwd`/resources/templates/digital_ocean/ubuntu.json --cluster `pwd`/resources/clusters/digital_ocean/cluster.json --kubernetes-version 1.13.4
```

2. All other tests (pulls in Kubeconfig from last run):

```
ginkgo -r -v -debug -failFast tests
```

Some things to note:

1. The only state required between the provision and test runs currently is the Kubeconfig. Ideally, I would like to keep it this way. We can do this by attaching any state to the cluster itself - e.g. add a label for the template ID (so it can be cleaned up as part of a cleanup suite), or add a ConfigMap, or...
2. We should think about being smarter with the Kubeconfig - maybe take a `--kubeconfig` arg so the environment variable isn't required (which can make running multiple tests at once locally annoying), maybe skip provisioning if the Kubeconfig already exists, ...

The most important TODOs are:

1. Decide on a real resource template strategy and add those templates
2. Update Dockerfile to actually run tests
3. Add CI configuration file (discussed with Nick - I would strongly prefer to use CircleCI, but we'll run out of build hours quickly)
4. Fix the scale test to test for the proper things
5. Add some more useful provision tests (check Coordinator API, etc)

Other TODOS are:

- Lots of refactoring to deduplicate code etc
- Add upgrade tests
- Add node pool add/remove tests
- Other tests

Closes https://github.com/containership/cloud-test/issues/50.
